src/web/routes.rs

```diff
@@ -0,0 +1,257 @@
+        if !pattern.ends_with('/') {
+            let pattern = format!("{}/", pattern);
+            self.get
+                .push((pattern.to_string(), Box::new(remove_slashes)));
+        }
+
+        // Register the prefix if it's not the home page and the first path component is not a
+        // pattern or a wildcard.
+        if pattern != "/" {
+            if let Some(first_component) = pattern.trim_matches('/').split('/').next() {
+                if !first_component.contains('*') && !first_component.starts_with(':') {
+                    self.page_prefixes.insert(first_component.to_string());
+                }
+            }
+        }
+    }
+
+    /// A rustdoc page is a page serving generated documentation. It's similar to a static
+    /// resource, but path prefixes are automatically blacklisted (see internal pages to learn more
+    /// about page prefixes).
+    fn rustdoc_page(&mut self, pattern: &str, handler: impl Handler) {
+        self.rustdoc_get
+            .push((pattern.to_string(), Box::new(handler)));
+    }
+}
+
+/// Simple Iron route that redirects to the same path without the trailing slash, for example
+/// /foo/bar/baz/ to /foo/bar/baz.
+fn remove_slashes(req: &mut iron::Request) -> iron::IronResult<iron::Response> {
+    let mut url: iron::url::Url = req.url.clone().into();
+    url.set_path(&url.path().trim_end_matches('/').to_string());
+    Ok(iron::Response::with((
+        iron::status::Found,
+        iron::modifiers::Redirect(iron::Url::parse(&url.to_string()).unwrap()),
+    )))
+}
+
+/// Iron Middleware that prevents requests to blacklisted prefixes.
+///
+/// In our application, a prefix is blacklisted if a docs.rs page exists below it. For example,
+/// since /releases/queue is a docs.rs page, /releases is a blacklisted prefix.
+///
+/// The middleware must be used for all the pages serving crates at the top level, to prevent a
+/// crate from putting their own content in an URL that's supposed to be used by docs.rs.
+pub(super) struct BlockBlacklistedPrefixes {
+    blacklist: HashSet<String>,
+    handler: Box<dyn Handler>,
+}
+
+impl BlockBlacklistedPrefixes {
+    pub(super) fn new(blacklist: HashSet<String>, handler: Box<dyn Handler>) -> Self {
+        Self { blacklist, handler }
+    }
+}
+
+impl Handler for BlockBlacklistedPrefixes {
+    fn handle(&self, req: &mut iron::Request) -> iron::IronResult<iron::Response> {
+        if let Some(prefix) = req.url.path().get(0) {
+            if self.blacklist.contains(*prefix) {
+                return Err(iron::IronError::new(
+                    super::error::Nope::CrateNotFound,
+                    iron::status::NotFound,
+                ));
+            }
+        }
+        self.handler.handle(req)
+    }
+}
+
+/// Automatically generate a Route ID from a pattern. Every non-alphanumeric character is replaced
+/// with `_`.
+fn calculate_id(pattern: &str) -> String {
```

Comment:
This is useful, the first mistake I always see new contributors making is duplicate route IDs so it's nice to avoid it automatically :)